### PR TITLE
Respect user enumeration settings in user status lists

### LINKED ISSUE #27122: user_status "last statuses" widget leaks account names
There should be an option to globally disable the "Last statuses" widget. It leaks account names, which might be the desired behaviour, but might as well be not.

On "semi-public" Nextcloud instances you usually don't want to expose other users to each other, which is also why e.g. `shareapi_allow_share_dialog_user_enumeration` can be disabled.

Currently it's only possible to disable user_status altogether. It would be nice if there was the option to keep user_status enabled but disable the "Last statuses" widget.

### PR DESCRIPTION
**Pull request by mejo-**
The functions to find user statuses listed didn't respect user
enumeration settings (`shareapi_allow_share_dialog_user_enumeration`
and `shareapi_restrict_user_enumeration_to_group` core app settings).
    
Fixes: #27122

### CODE REVIEW

**Comment by mejo-** [comment]
As mentioned in [issue #27122](https://github.com/nextcloud/server/issues/27122#issuecomment-873665680), I'd prefer to remove the route `Statuses#findAll` and `statusService->findAll()` altogether. But in order to be the least intrusive, for now this PR only implements the second part of my suggested fix:

> Make `statusService->findAllRecentStatusChanges` return an empty array if config option `shareapi_allow_share_dialog_user_enumeration` is disabled.

Please let me know what you think about removing the `findAll` routes and functions.

**Comment by nickvergessen** [comment]
From my POV this is not enumeration. You can not start to type a name and find the users. but of course since the limit and offset are controlled, you can still create a user list, so sounds good.

But if `shareapi_restrict_user_enumeration_to_phone` or `shareapi_restrict_user_enumeration_to_group` is enabled we should return those users I think. Maybe https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/private/Collaboration/Collaborators/UserPlugin.php#L99-L134 can give you some inspiration

**Comment by mejo-** [comment]
> From my POV this is not enumeration. You can not start to type a name and find the users. but of course since the limit and offset are controlled, you can still create a user list, so sounds good.

You're right, strictly speaking it's not enumeration. But with the `Statuses#findAll` route, you can simply get a list of *all* users on the instance, which is even worse than enumeration :wink: And as you said, by controlling limits and offsets, it's possible to get all users even with the `Statuses#find` route.

**Comment by mejo-** [comment]
> But if `shareapi_restrict_user_enumeration_to_phone` or `shareapi_restrict_user_enumeration_to_group` is enabled we should return those users I think.

If I understand it correctly, then this would be further limitations to if `shareapi_allow_share_dialog_user_enumeration` is **enabled**, right? If `shareapi_allow_share_dialog_user_enumeration` is **disabled**, then it's not possible to select those two options you mentioned. So my understanding is that they're there to limit the scope of user enumeration in case it's enabled in general.

So far, this PR is about fixing the user_status API to respect admins choice to disable user enumeration altogether.

Would you prefer to implement support for `shareapi_restrict_user_enumeration_to_phone` and `shareapi_restrict_user_enumeration_to_group` in this PR as well? If you want me to, I could certainly look into it. But it would increase scope and complexibility of this PR a lot :wink:

**Comment by nickvergessen** [comment]
> So my understanding is that they're there to limit the scope of user enumeration in case it's enabled in general.

Yeah, you are right. My head tricked me thinking it was the other way around.

> Would you prefer to implement support for shareapi_restrict_user_enumeration_to_phone and shareapi_restrict_user_enumeration_to_group in this PR as well? If you want me to, I could certainly look into it. But it would increase scope and complexibility of this PR a lot wink

Yes, because either we say the setting does influence it, then the further settings should too, or not then we don't need this PR or a follow up.

**Comment by mejo-** [comment]
So I took a deeper look now and changed the logic to respect both `shareapi_allow_share_dialog_user_enumeration` and `shareapi_restrict_user_enumeration_to_group` as suggested by @nickvergessen :blush: 

As far as I understand, `shareapi_restrict_user_enumeration_to_phone` is not relevant here as it's only about limiting partial matches for particular user names.

**Comment by mejo-** [comment]
Now all tests are passing again. Sorry, the last days CI had some hickups and I didn't spot that the drone check didn't pass due to errors in the unit tests.

**Comment by mejo-** [comment]
Mh, any chance to get a review here :blush:

**Comment by nickvergessen** [comment]
> As far as I understand, shareapi_restrict_user_enumeration_to_phone is not relevant here as it's only about limiting partial matches for particular user names.

Well so it should also be respected if possible.

Basically if:
- shareapi_allow_share_dialog_user_enumeration: yes
- shareapi_restrict_user_enumeration_to_group: no
- shareapi_restrict_user_enumeration_to_phone: yes

you should only see users which you have the phone number matched.
But I would be okay to ignore it for now and fix this in a follow up as it increases complexity a lot.

**Comment by mejo-** [comment]
>  But I would be okay to ignore it for now and fix this in a follow up as it increases complexity a lot.

@nickvergessen thanks for your feedback :blush: 

So are you ok with the PR as is or do you see the need for further changes/improvements?

File: apps/user_status/lib/Db/UserStatusMapper.php [hunk 1]
```diff
@@ -116,13 +116,51 @@ public function findByUserId(string $userId):UserStatus {
...
 			->select('*')
 			->from($this->tableName)
 			->where($qb->expr()->in('user_id', $qb->createNamedParameter($userIds, IQueryBuilder::PARAM_STR_ARRAY)));
 
+		if ($limit !== null) {
+			$qb->setMaxResults($limit);
+		}
+		if ($offset !== null) {
+			$qb->setFirstResult($offset);
+		}
```
Comment (nickvergessen) [inline apps/user_status/lib/Db/UserStatusMapper.php:131]: when you paginate, you should explicitly set an `orderBy` to make sure all DBs behave the same

File: apps/user_status/lib/Service/StatusService.php [hunk 2]
```diff
@@ -328,6 +379,26 @@ public function removeUserStatus(string $userId): bool {
...
+			$usersInGroup = $this->groupManager->displayNamesInGroup($userGroupId);
+			foreach ($usersInGroup as $userId => $displayName) {
+				$userId = (string)$userId;
+				if (!in_array($userId, $usersByGroups, true)) {
+					$usersByGroups[] = $userId;
+				}
+			}
+		}
+
+		return $usersByGroups;
```
Comment (nickvergessen) [inline apps/user_status/lib/Service/StatusService.php:399]: I think this does not scale. We need to do it in another way. In worst case someone is in hundreds of (automated) groups, you run hundreds of queries to generate the user list to then later know that they don't even have a status set or something. Can't we do it similar to `lib/private/Collaboration/Collaborators/UserPlugin.php` and first get the recent user statuses and then filter out the results?
Basically get the status of 100 people, then filter out, if less than the requested users remain retry
